File: Java/EX11.java

// Classe abstrata Funcionario
abstract class Funcionario {
    protected String nome;

    // Método abstrato para calcular salário
    public abstract double calcularSalario();

    // Método para obter o nome do funcionário
    public String getNome() {
        return nome;
    }
}

// Classe derivada FuncionarioHorista
class FuncionarioHorista extends Funcionario {
    private int horasTrabalhadas;
    private double valorHora;

    // Construtor
    public FuncionarioHorista(String nome, int horasTrabalhadas, double valorHora) {
        this.nome = nome;
        this.horasTrabalhadas = horasTrabalhadas;
        this.valorHora = valorHora;
    }

    // Implementação do método calcularSalario
    @Override
    public double calcularSalario() {
        return horasTrabalhadas * valorHora;
    }
}

// Classe derivada FuncionarioAssalariado
class FuncionarioAssalariado extends Funcionario {
    private double salarioMensal;

    // Construtor
    public FuncionarioAssalariado(String nome, double salarioMensal) {
        this.nome = nome;
        this.salarioMensal = salarioMensal;
    }

    // Implementação do método calcularSalario
    @Override
    public double calcularSalario() {
        return salarioMensal;
    }
}

// Classe principal para teste
class Main6 {
    public static void main(String[] args) {
        // Criando um funcionário horista
        Funcionario funcionarioHorista = new FuncionarioHorista("Carlos", 203, 15);
        System.out.printf("Salário de %s: R$%.2f%n", funcionarioHorista.getNome(), funcionarioHorista.calcularSalario());

        // Criando um funcionário assalariado
        Funcionario funcionarioAssalariado = new FuncionarioAssalariado("Melissa", 3000);
        System.out.printf("Salário de %s: R$%.2f%n", funcionarioAssalariado.getNome(), funcionarioAssalariado.calcularSalario());
    }
}

File: Go/EX11.go
package main

import (
	"fmt"
)

// Definição da interface Funcionario
type Funcionario interface {
	CalcularSalario() float64
	GetNome() string
}

// Estrutura FuncionarioHorista
type FuncionarioHorista struct {
	nome             string
	horasTrabalhadas int
	valorHora        float64
}

// Construtor para FuncionarioHorista
func NewFuncionarioHorista(nome string, horasTrabalhadas int, valorHora float64) *FuncionarioHorista {
	return &FuncionarioHorista{nome: nome, horasTrabalhadas: horasTrabalhadas, valorHora: valorHora}
}

// Implementação do método CalcularSalario
func (f *FuncionarioHorista) CalcularSalario() float64 {
	return float64(f.horasTrabalhadas) * f.valorHora
}

// Implementação do método GetNome
func (f *FuncionarioHorista) GetNome() string {
	return f.nome
}

// Estrutura FuncionarioAssalariado
type FuncionarioAssalariado struct {
	nome          string
	salarioMensal float64
}

// Construtor para FuncionarioAssalariado
func NewFuncionarioAssalariado(nome string, salarioMensal float64) *FuncionarioAssalariado {
	return &FuncionarioAssalariado{nome: nome, salarioMensal: salarioMensal}
}

// Implementação do método CalcularSalario
func (f *FuncionarioAssalariado) CalcularSalario() float64 {
	return f.salarioMensal
}

// Implementação do método GetNome
func (f *FuncionarioAssalariado) GetNome() string {
	return f.nome
}

// Função principal para teste
func main() {
	// Criando um funcionário horista
	funcionarioHorista := NewFuncionarioHorista("Carlos", 203, 15)
	fmt.Printf("Salário de %s: R$%.2f\n", funcionarioHorista.GetNome(), funcionarioHorista.CalcularSalario())

	// Criando um funcionário assalariado
	funcionarioAssalariado := NewFuncionarioAssalariado("Melissa", 3000)
	fmt.Printf("Salário de %s: R$%.2f\n", funcionarioAssalariado.GetNome(), funcionarioAssalariado.CalcularSalario())
}
